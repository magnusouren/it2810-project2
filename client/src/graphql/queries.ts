import { gql } from '@apollo/client';

export const GET_MOVIES = gql`
  query getMovies($page: Int!) {
    getMovies(page: $page) {
      _id
      genre_ids
      poster_path
      title
    }
  }
`;

export const GET_MOVIES_BY_GENRE = gql`
  query getMoviesByGenre($page: Int!, $genreId: Int!) {
    getMoviesByGenre(page: $page, genreId: $genreId) {
      _id
      genre_ids
      poster_path
      title
    }
    getMovieCountByGenre(genreId: $genreId)
  }
`;

export const GET_MOVIES_BY_TITLE_AZ = gql`
  query getMoviesByTitleAZ($page: Int!, $order: String!, $genreId: Int) {
    getMoviesByTitleAZ(page: $page, order: $order, genreId: $genreId) {
      _id
      genre_ids
      poster_path
      title
    }
    getMovieCountByGenre(genreId: $genreId)
  }
`;

export const GET_MOVIES_BY_RATING = gql`
  query getMoviesByRating($page: Int!, $order: String!, $genreId: Int) {
    getMoviesByRating(page: $page, order: $order, genreId: $genreId) {
      _id
      genre_ids
      poster_path
      title
    }
    getMovieCountByGenre(genreId: $genreId)
  }
`;

export const GET_MOVIES_BY_TITLE = gql`
  query getMoviesByTitle($title: String!, $limit: Int!) {
    getMoviesByTitle(title: $title, limit: $limit) {
      _id
      title
    }
  }
`;

export const IS_IN_WATCHLIST = gql`
  query Query($userId: String!, $movieId: Int!) {
    movieIsInWatchlist(userID: $userId, movieID: $movieId)
  }
`;

export const REMOVE_MOVIE_FROM_WATCHLIST = gql`
  mutation Mutation($userId: String!, $movieId: Int!) {
    removeMovieFromWatchlist(userID: $userId, movieID: $movieId) {
      userID
      movies {
        _id
      }
    }
  }
`;

export const ADD_MOVIE_TO_WATCHLIST = gql`
  mutation Mutation($userId: String!, $movieId: Int!) {
    addMovieToWatchlist(userID: $userId, movieID: $movieId) {
      userID
      movies {
        _id
      }
    }
  }
`;

export const GET_WATCHLIST_BY_USER_ID = gql`
  query Query($userId: String!, $page: Int!) {
    getWatchlistByUserID(userID: $userId, page: $page) {
      userID
      movies {
        _id
        title
        genre_ids
        poster_path
      }
    }
    getWatchlistCountByUserID(userID: $userId)
  }
`;

export const GET_MOVIE_RATING_WITH_USERID = gql`
  query getMovieRatingWithUserID($userID: String!, $movieID: Int!) {
    getMovieRatingWithUserID(userID: $userID, movieID: $movieID) {
      rating
    }
  }
`;

export const ADD_RATING = gql`
  mutation addRating($userID: String!, $movieID: Int!, $rating: Float!) {
    addRating(userID: $userID, movieID: $movieID, rating: $rating) {
      rating
    }
  }
`;

export const GET_FILTER = gql`
  query FilterData {
    filter @client {
      alphabeticalSort
      genre
      ratingSort
      page
    }
  }
`;

export const determineQueryAndVariables = (
  page: number,
  genre: string,
  alphabeticalSort: string,
  ratingSort: string,
) => {
  const baseVariables = { page: page };

  if (genre) {
    if (alphabeticalSort) {
      return {
        query: GET_MOVIES_BY_TITLE_AZ,
        variables: { ...baseVariables, genreId: genre, order: alphabeticalSort },
      };
    }
    if (ratingSort) {
      return {
        query: GET_MOVIES_BY_RATING,
        variables: { ...baseVariables, genreId: genre, order: ratingSort },
      };
    }
    return {
      query: GET_MOVIES_BY_GENRE,
      variables: { ...baseVariables, genreId: genre },
    };
  }

  if (alphabeticalSort) {
    return {
      query: GET_MOVIES_BY_TITLE_AZ,
      variables: { ...baseVariables, order: alphabeticalSort },
    };
  }
  if (ratingSort) {
    console.log('ratingSort', ratingSort);
    return {
      query: GET_MOVIES_BY_RATING,
      variables: { ...baseVariables, order: ratingSort },
    };
  }
  return {
    query: GET_MOVIES,
    variables: baseVariables,
  };
};
