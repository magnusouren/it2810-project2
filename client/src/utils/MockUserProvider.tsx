import { FC, ReactNode, useState } from 'react';
import { vi } from 'vitest';

// Importer UserContext
import { UserContext } from '../context/UserContext'; // Importer UserContext
import { User } from '../types';

interface MockUserProviderProps {
  children: ReactNode;
  mockUser?: User;
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
}

const MockUserProvider: FC<MockUserProviderProps> = ({
  children,
  mockUser,
  loginMock = vi.fn(),
  logoutMock = vi.fn(),
  deleteUserMock = vi.fn(),
  toggleDarkModeMock = vi.fn(),
}) => {
  const [user, setUser] = useState<User | undefined>(mockUser);

  const login = loginMock;
  const logout = logoutMock;
  const deleteUser = deleteUserMock;
  const toggleDarkMode = toggleDarkModeMock;

  const value = {
    user,
    setUser,
    login,
    logout,
    deleteUser,
    darkMode: false,
    toggleDarkMode,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default MockUserProvider;
