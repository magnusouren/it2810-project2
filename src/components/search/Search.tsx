import SearchIcon from '@mui/icons-material/Search';
import { InputAdornment, TextField } from '@mui/material';

import styles from './Search.module.scss';

const Search: React.FC = () => {
  // TODO: Search logic
  return (
    <TextField
      variant='outlined'
      placeholder='Search..'
      className={styles.searchBar}
      InputProps={{
        startAdornment: (
          <InputAdornment position='start'>
            <SearchIcon />
          </InputAdornment>
        ),
      }}
    />
  );
};
export default Search;
