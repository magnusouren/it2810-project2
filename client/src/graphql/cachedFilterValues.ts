import { client } from '../main';
import { Sort } from '../types';
import { GET_FILTER } from './queries';

type FilterResponse = {
  filter: Filter | null;
};

type Filter = {
  sort: Sort;
  genre: string;
  page: number;
};

/**
 *
 * Gets the cached filter values.
 *
 * @returns {Filter}
 */
export const getCachedFilterValues = () => {
  const cache = client.cache;
  const filterResponse = cache.readQuery<FilterResponse>({
    query: GET_FILTER,
  });
  const filter: Filter = filterResponse?.filter || { sort: '', genre: '', page: 1 };

  return filter;
};

/**
 *
 * Sets the cached filter values.
 *
 * @param {Filter} filter
 */
export const setCachedFilterValues = ({ sort, genre, page }: Filter) => {
  const cache = client.cache;

  cache.writeQuery({
    query: GET_FILTER,
    data: {
      filter: {
        sort: sort,
        genre: genre,
        page: page,
      },
    },
  });
};
