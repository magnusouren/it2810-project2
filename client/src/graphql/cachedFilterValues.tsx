import { client } from '../main';
import { AlphabeticalSort, RatingSort } from '../types';
import { GET_FILTER } from './queries';

type FilterResponse = {
  filter: Filter | null;
};

type Filter = {
  alphabeticalSort: AlphabeticalSort;
  genre: string;
  ratingSort: RatingSort;
  page: number;
};

/**
 *
 * Gets the cached filter values.
 *
 * @returns {Filter}
 */
export const getCachedFilterValues = () => {
  const cache = client.cache;
  const filterResponse = cache.readQuery<FilterResponse>({
    query: GET_FILTER,
  });
  const filter: Filter = filterResponse?.filter || { alphabeticalSort: '', genre: '', ratingSort: '', page: 1 };

  return filter;
};

/**
 *
 * Sets the cached filter values.
 *
 * @param {Filter} filter
 */
export const setCachedFilterValues = ({ alphabeticalSort, genre, ratingSort, page }: Filter) => {
  const cache = client.cache;

  cache.writeQuery({
    query: GET_FILTER,
    data: {
      filter: {
        alphabeticalSort: alphabeticalSort,
        genre: genre,
        ratingSort: ratingSort,
        page: page,
      },
    },
  });
};
