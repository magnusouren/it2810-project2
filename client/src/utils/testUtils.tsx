import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { User } from '../types';
import MockUserProvider from './MockUserProvider';

const link = new HttpLink({
  uri: 'http://localhost:4000', // Endepunktet til din GraphQL server
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

type ProviderMocks = {
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
};

interface ProviderProps {
  child: React.ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mocks?: any[];
  mockUser?: User;
  providerMocks?: ProviderMocks;
}

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithProviders = ({ child, mockUser, mocks, providerMocks }: ProviderProps) =>
  render(
    <MockUserProvider mockUser={mockUser} {...providerMocks}>
      <ApolloProvider client={client}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <BrowserRouter>{child}</BrowserRouter>
        </MockedProvider>
      </ApolloProvider>
    </MockUserProvider>,
  );
