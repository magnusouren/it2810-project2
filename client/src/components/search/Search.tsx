import { useQuery } from '@apollo/client';
import SearchIcon from '@mui/icons-material/Search';
import { InputAdornment, TextField } from '@mui/material';
import React, { ChangeEvent, useCallback, useEffect, useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import { GET_MOVIES_BY_TITLE } from '../../graphql/queries';
import styles from './Search.module.scss';

/**
 *
 * Search component.
 * Displays a search bar that filters movies by title.
 *
 * @returns {React.JSX.Element}
 *
 */
interface SearchResult {
  title: string;
  _id: string;
}

export const Search: React.FC = () => {
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [debouncedSearchTerm, setDebouncedSearchTerm] = useState<string>('');
  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const [limit, setLimit] = useState<number>(20);
  const [tabIndex, setTabIndex] = useState<number>(1);
  const navigate = useNavigate();
  const dropdownRef = useRef(null);

  // Use useEffect to debounce search term after user input to prevent too many queries
  useEffect(() => {
    const timerId = setTimeout(() => {
      setDebouncedSearchTerm(searchTerm.trim()); // Trim when updating debounced term
    }, 300); // 300ms delay

    return () => {
      clearTimeout(timerId);
    };
  }, [searchTerm]);

  // Use useQuery to fetch movies by title
  const { data } = useQuery(GET_MOVIES_BY_TITLE, {
    variables: { title: debouncedSearchTerm, limit: limit },
    skip: !debouncedSearchTerm, // Skips the query if searchTerm is empty
  });

  // Use useCallback to prevent to many rerenders
  const handleInputChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(e.target.value);
    setIsDropdownOpen(true);
    setLimit(20);
    setTabIndex(1);
  }, []);

  // Use useCallback to prevent to many rerenders
  const handleMovieSelect = useCallback(
    (movie: SearchResult) => {
      navigate(`/movie/${movie._id}`);
      setIsDropdownOpen(false);
      setSearchTerm('');
    },
    [navigate],
  );

  // Use useEffect to close dropdown when user clicks outside of it
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !(dropdownRef.current as Node).contains(event.target as Node)) {
        setIsDropdownOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  const loadMoreMovies = () => {
    setLimit(limit + 20);
  };

  const handleKeyDown = (e: React.KeyboardEvent, movie: SearchResult) => {
    if (e.key === 'Tab') {
      if (e.shiftKey) {
        setTabIndex(tabIndex - 1);
        return;
      }
      setTabIndex(tabIndex + 1);
      if (data?.getMoviesByTitle.length < limit && tabIndex === data?.getMoviesByTitle.length) {
        setIsDropdownOpen(false);
      }
    }
    if (e.key === 'Enter') {
      handleMovieSelect(movie);
    }
    if (e.key === 'Escape') {
      setIsDropdownOpen(false);
      setSearchTerm('');
    }
  };

  return (
    <div className={styles.searchBar} data-testid='searchbar-container'>
      <TextField
        id='searchBar'
        variant='outlined'
        placeholder='Search by title...'
        value={searchTerm}
        onChange={handleInputChange}
        onKeyDown={(e) => handleKeyDown(e, data?.getMoviesByTitle[0])}
        className={styles.searchInput}
        InputProps={{
          startAdornment: (
            <InputAdornment position='start'>
              <SearchIcon />
            </InputAdornment>
          ),
        }}
      />

      {/* Dropdown of filtered movies */}
      {data?.getMoviesByTitle.length === 0 && isDropdownOpen && data && searchTerm.length > 0 && (
        <ul className={styles.dropdown} ref={dropdownRef}>
          <li id={styles.notFound}>No results found for {`"${searchTerm}"`}</li>
        </ul>
      )}
      {data?.getMoviesByTitle.length > 0 && isDropdownOpen && (
        <ul className={styles.dropdown} ref={dropdownRef} tabIndex={0} aria-label='Search result for movie title'>
          {data.getMoviesByTitle.map((movie: SearchResult) => (
            <li
              key={movie._id}
              className={styles.movie}
              tabIndex={0}
              onClick={() => handleMovieSelect(movie)}
              onKeyDown={(e) => handleKeyDown(e, movie)}
              aria-label={`Movie: ${movie.title}`}
            >
              {movie.title}
            </li>
          ))}
          {data?.getMoviesByTitle.length % 20 === 0 && (
            <li
              className={styles.loadMore}
              onClick={loadMoreMovies}
              tabIndex={0}
              onKeyDown={(e) => {
                if (e.key === 'Enter') {
                  loadMoreMovies();
                  setTabIndex(1);
                } else if (e.key === 'Tab' && !e.shiftKey) {
                  setIsDropdownOpen(false);
                } else if (e.key == 'Tab' && e.shiftKey) {
                  setTabIndex(tabIndex - 1);
                }
              }}
            >
              Load more movies...
            </li>
          )}
        </ul>
      )}
    </div>
  );
};
