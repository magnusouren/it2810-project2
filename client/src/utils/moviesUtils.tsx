import { GET_MOVIES, GET_MOVIES_BY_GENRE, GET_MOVIES_BY_RATING, GET_MOVIES_BY_TITLE_AZ } from '../graphql/queries';

export const determineQueryAndVariables = (page: number, genre: string, sort: string, userId: string) => {
  const baseVariables = { page: page, userId: userId };
  const alphabeticalSort = sort === 'a-z' || sort === 'z-a';
  const ratingSort = sort === 'h-l' || sort === 'l-h';

  if (genre) {
    if (alphabeticalSort) {
      return {
        query: GET_MOVIES_BY_TITLE_AZ,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    if (ratingSort) {
      return {
        query: GET_MOVIES_BY_RATING,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    return {
      query: GET_MOVIES_BY_GENRE,
      variables: { ...baseVariables, genreId: genre },
    };
  }

  if (alphabeticalSort) {
    return {
      query: GET_MOVIES_BY_TITLE_AZ,
      variables: { ...baseVariables, order: sort },
    };
  }
  if (ratingSort) {
    return {
      query: GET_MOVIES_BY_RATING,
      variables: { ...baseVariables, order: sort },
    };
  }
  return {
    query: GET_MOVIES,
    variables: { ...baseVariables },
  };
};
