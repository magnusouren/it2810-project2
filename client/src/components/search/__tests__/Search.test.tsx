import { fireEvent, screen, waitFor } from '@testing-library/react';

import { GET_MOVIES_BY_TITLE } from '../../../graphql/queries';
import { renderWithProviders } from '../../../utils/testUtils';
import Search from '../Search';

describe('Search', () => {
  it('Should render input field', () => {
    renderWithProviders(<Search />);
    expect(screen.getByRole('textbox')).toBeDefined();
  });

  it('Should match snapshot', () => {
    const { container } = renderWithProviders(<Search />);
    expect(container).toMatchSnapshot();
  });

  it('Should show dropdown when input field is focused and has value', async () => {
    const mocks = [
      {
        request: {
          query: GET_MOVIES_BY_TITLE,
          variables: {
            title: 'foo',
            limit: 20,
          },
        },
        result: {
          data: {
            getMoviesByTitle: [
              {
                _id: '1',
                title: 'Foo',
              },
            ],
          },
        },
      },
    ];

    renderWithProviders(<Search />, mocks);
    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'foo' } });
    fireEvent.focus(screen.getByRole('textbox'));

    await waitFor(() => {
      expect(screen.getByRole('list')).toBeDefined();
    });
  });
});
