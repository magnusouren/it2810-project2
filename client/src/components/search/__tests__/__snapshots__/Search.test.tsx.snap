// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > Should match snapshot 1`] = `
<div>
  <div
    class="_searchBar_6ac984"
  >
    <div
      class="MuiFormControl-root MuiTextField-root _searchInput_6ac984 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
          id="searchBar"
          placeholder="Search by title..."
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
